Model designer update

jonathanlurie opened the pull request:
# Styling the Model Designer 🧁
## Description

This work is a follow up on @ValentinBuira 's drag and drop support in Model Designer https://github.com/qgis/QGIS/pull/60664

It essentially adds some visual feedback to it. Here is a global view:
![Screenshot from 2025-05-07 15-32-58](https://github.com/user-attachments/assets/f4b23815-fae3-422c-b9c3-607b201e0622)

Now let's see some detailed view of a processing box:
![Screenshot from 2025-05-08 10-14-43](https://github.com/user-attachments/assets/e1b4b429-a5ef-4c7c-9655-3e78407d70b3)

And of a parameter box:
![Screenshot from 2025-05-08 10-15-06](https://github.com/user-attachments/assets/fcb2dea6-f5f7-445e-9704-fe16d9d03d46)


- Bullets are now color coded by type of data
  - blue rgb(34, 157, 214) --> numerical types (numbers, matrices, vectors, distances, durations, etc.)
  - green rgb(51, 201, 28) --> boolean
  - kaki (greenish yellow) rgb(180, 180, 0) --> vector geometries and vector layers. This includes meshes, point clouds, extents, vector tiles and anything that contains some sort of geometry
  - turquoise rgb(0, 180, 180) --> raster images and raster layers
  - purple rgb(128, 68, 201) --> enums
  - slate blueish rgb(100, 100, 255) --> everything text: strings, date, CRS, HTML
  - dark gray rgb(80, 80, 80) --> everything file system
  - mid gray rgb(128, 128, 128) --> the rest that is difficult to find a category for (mode datails here https://github.com/camptocamp/QGIS/blob/model-designer-update/src/gui/processing/models/qgsmodelgraphicitem.cpp#L232)
- rubber bands have the same color coding as bullets

- Bullets are just outlined when holding the default value, or filled if configured with a non-default value (or a linked to a parameter box)
- The label of each param is followed by its configured value. It used to update only after the whole mode was ran, now it updates whenever the used updates the value or plugs a parameter box
- values that are enums are displaying the human-friendly label of the corresponding enum value (and not just the index of it)
- values that are coming from a plugged box clearly display `value from "<NAME OF THIS BOX>"` (when a parameter) or `"<OUTPUT NAME> from <PROCESSING NODE NAME>" ` 
- parameter boxes also display the type and value on their "out" section when unfolded

Note: color choice can obviously be discussed and when I picked those it was a mix of randomness and looking-alrightness. I am particularly interested in feedback from colorblind users, so that we can represent types that are very different with colors that are very different for colorblinds too.

jonathanlurie committed:
adding model designer style customization

github-actions[bot] committed:
auto-fix pre-commit issues

ValentinBuira commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -1140,9 +1238,60 @@ QString QgsModelChildAlgorithmGraphicItem::linkPointText( Qt::Edge edge, int ind
+              paramValueAsStr = QStringLiteral( ": %1" ).arg( paramValue.toString() );
+
+              // In case of an enum, we want to display the label of the enum value (and not just its index as an int)
+              if ( param->type() == QgsProcessingParameterEnum::typeName() )
```
What about `QgsProcessingParameterCrs` or  `QgsProcessingParameterFieldMapping` ? I think there are holes in the racket. 

take a look as how as `QgsProcessingParameterDefinition::asPythonString`  and `QgsProcessingParameterDefinition::valueAsString` and subclasses for a more generic approach

QgsProcessingParameterDefinition::asPythonString is used when exporting the model to a python script

![image](https://github.com/user-attachments/assets/02774271-8fcc-4658-b14d-60dfbb6624e2)

jonathanlurie replied:
Good point! The issue with `QgsProcessingParameterDefinition::asPythonString()` and `QgsProcessingParameterDefinition::valueAsString()` is that they need to be passed a context (QgsProcessingContext), which I cannot provide from `QgsModelChildAlgorithmGraphicItem::linkPointText` (as far as I know). This is the reason I am using `child->parameterSources()[0].staticValue()`  instead. I spent quite some time to look for an alternative or a way to access the context (without bridging gaps that shouldn't be bridged) but without success.

jonathanlurie replied:
Update: i just noticed that the context is actually not necessary for child classes of QgsProcessingParameterDefinition for methods such as valueAsPythonString 🤦

So yeah, i'll do something in the same vibe for getting a user friendly string version of the param value.

ValentinBuira reviewed (changes requested):
IMO there is still some modifications to make this merge-ready. But this already make the modeler WAY more appealing !

jonathanlurie committed:
Merge branch 'master' into model-designer-update

jonathanlurie committed:
Added color names

jonathanlurie committed:
testing with better line jumps 2

jonathanlurie committed:
formatting

jonathanlurie commented on src/core/processing/models/qgsprocessingmodelchildparametersource.h:
```diff
@@ -230,6 +230,11 @@ class CORE_EXPORT QgsProcessingModelChildParameterSource
+    /**
+     * Returns the type of source
+     */
+    Qgis::ProcessingModelChildParameterSource getSourceType() { return mSource; }
```
indeed, how could i missed that??

jonathanlurie committed:
now using virtual method for getting param data type

nyalldawson commented:
@jonathanlurie 

Before diving into the code I'd like to discuss a few high-level considerations from the UI changes:

> Bullets are now color coded by type of data
> vector geometries and vector layers. This includes meshes, point clouds, extents, vector tiles and anything that contains some sort of geometry

Point clouds, mesh and vector tile layers **aren't** vector layers. They won't be compatible with vector layer parameters, or with each other. I'd suggest using a different color for these "other" layer types for now. (ie one color for vector layers/vector features, one for raster layers, and one for all other layers)

> rubber bands have the same color coding as bullets

Given that joins are tolerant and that you can eg join a numeric output to a string input, how is this handled? (My gut feeling is that we should always use the coloring of the output, as that's representative of the actual value "flowing" along the link)

> The label of each param is followed by its configured value. It used to update only after the whole mode was ran, now it updates whenever the used updates the value or plugs a parameter box

Sounds good!

> values that are enums are displaying the human-friendly label of the corresponding enum value (and not just the index of it)

Good move!

> values that are coming from a plugged box clearly display "value from <NAME OF THIS BOX>" (when a parameter) or "<OUTPUT NAME> from <PROCESSING NODE NAME>" 

This change concerns me, because it's resulting in very long strings. You can see in your screenshots that you've had to greatly widen the boxes in order to actually fit this text. 

For the values linked to parameters I'd drop the "value from ..." prefix and just use "INPUT: param name" (ie. "Keep disjoint parts separate: Some bool value" in your screenshot). Or even better "INPUT: param name (last used value)" after the model has been run (i.e. Keep disjoint parts separate: Some bool value (FALSE)").

For the values linked to alg outputs I'd just go with the output name "Input layer: fixed geometries" and rely on the rubber band as the visual link to the algorithm. Or alternatively, responsively adapt the text to the actual available width, so we'd use "input layer: fixed geometries" unless the box is wide enough to show the full "input layer: fixed geometries from fix geometries" text.

For 
> parameter boxes also display the type and value on their "out" section when unfolded

Sounds good!

ValentinBuira commented:
> Given that joins are tolerant and that you can eg join a numeric output to a string input, how is this handled? (My gut feeling is that we should always use the coloring of the output, as that's representative of the actual value "flowing" along the link)

Blender also have implicit conversion and no direct cast and they solve the issue by applying a gradiant along the link

for example 
![Screenshot From 2025-05-13 10-45-00](https://github.com/user-attachments/assets/77932495-38f6-47b8-b0b6-525b2192a529)

Maybe we could do someting similar with a `QLinearGradient` and sampling both start and end socket color ?

ValentinBuira commented:
> This change concerns me, because it's resulting in very long strings. You can see in your screenshots that you've had to greatly widen the boxes in order to actually fit this text.

This is outside the scope of this PR, but when a component is out of the current view.  I would love to see a brief description of the other end of the link at the edge on the viewed content

 I don't think I ever saw this in any node based application I  used so far.

For example on the left with the `Output from Difference`

![image psd(8)](https://github.com/user-attachments/assets/eed8cdbd-7083-40d4-8e8a-22aa0ce32b25)

jonathanlurie committed:
simplified button positioning

jonathanlurie committed:
removed var used only once

jonathanlurie committed:
return empty string to signal error

jonathanlurie committed:
removed unnecessary else returns

jonathanlurie commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -197,4 +209,194 @@ void QgsModelDesignerSocketGraphicItem::paint( QPainter *painter, const QStyleOp
+{
+  // Numerical types
+  if (
+    dataType == QgsProcessingParameterMatrix::typeName() || dataType == QgsProcessingParameterNumber::typeName() || dataType == QgsProcessingParameterRange::typeName() || dataType == QgsProcessingParameterColor::typeName() || dataType == QgsProcessingOutputNumber::typeName() || dataType == QgsProcessingParameterDistance::typeName() || dataType == QgsProcessingParameterDuration::typeName() || data ...
```
Indeed, that would be a much better design! The colorCode() method could even be static in this case

jonathanlurie replied:
Good point! I just made a refactoring of how the color logic works in the last commit

jonathanlurie committed:
refectored param color logic

jonathanlurie committed:
moved virtual method body to cpp file

jonathanlurie committed:
moved virtual method body to cpp file 2

jonathanlurie committed:
simplified getting socket colors

jonathanlurie committed:
moving the socket with a small offset to better align

jonathanlurie commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -180,12 +185,19 @@ QgsModelDesignerSocketGraphicItem::QgsModelDesignerSocketGraphicItem( QgsModelCo
   painter->setRenderHint( QPainter::Antialiasing );
 
-  constexpr float DISPLAY_SIZE = 3.2;
+  constexpr float DISPLAY_SIZE = 4;
```
Oh I havnt noticed (until i zoomedin). Fixed now

jonathanlurie committed:
Adjusted start socket position and style

jonathanlurie committed:
merging switch cases

jonathanlurie committed:
output socket now are always filled

jonathanlurie commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -180,12 +185,19 @@ QgsModelDesignerSocketGraphicItem::QgsModelDesignerSocketGraphicItem( QgsModelCo
-  painter->setBrush( QBrush( QColor( 0, 0, 0, mHoverState ? 200 : 33 ), Qt::SolidPattern ) );
+  QColor outlineColor = getColor();
+  QColor fillColor = QColor( outlineColor );
+  fillColor.setAlpha( isDefaultParamValue() ? 30 : 255 );
```
Good point!  
The decision I took about the input sockets is that they would only be filled if the value they carry is different from the default value, whether it's because it was set or because a param box is plugged to it. I figured it was more relevant than filling the socket only when some external box is plugged to it because we already have a visual feedback about plugged/unplugged with the bezier.  

Up to now I kind of didn't really take a decision about whether or not the output socket should look and behave different than the input sockets. But now it's quite clear that it should indeed be a bit different. What I am suggesting (and pushed as part of this PR) is that output sockets are always filled (for both param and processing boxes) because they will always bear a (non-default) value

jonathanlurie commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -1140,9 +1238,60 @@ QString QgsModelChildAlgorithmGraphicItem::linkPointText( Qt::Edge edge, int ind
+        QgsProcessingModelChildParameterSources paramSources = child->parameterSources().value( name );
+        QString paramValueAsStr = "";
+
+        if ( paramSources.size() > 0 )
```
It feels a bit risky to read `paramSources[0]` without checking, that's an easily avoided segfault

jonathanlurie commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -1115,10 +1217,6 @@ QString QgsModelChildAlgorithmGraphicItem::linkPointText( Qt::Edge edge, int ind
         QString title = output->description();
-        if ( outputs.contains( output->name() ) )
-        {
-          title += QStringLiteral( ": %1" ).arg( outputs.value( output->name() ).toString() );
```
Ok. Do we want to have it back? It feels a bit cryptic. Also, keeping info from a previous run would prevent the field from being update with the last configured value that's going to be used in the next run.

jonathanlurie commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -842,13 +868,44 @@ QString QgsModelParameterGraphicItem::linkPointText( Qt::Edge, int index ) const
 }
 
+
+QString QgsModelParameterGraphicItem::getLinkedParamDataType( Qt::Edge edge, int index )
```
removed

jonathanlurie committed:
WIP tailored prompt for param type

jonathanlurie committed:
Merge branch 'master' into model-designer-update

jonathanlurie committed:
Merge branch 'master' into model-designer-update

jonathanlurie committed:
Update color groups

jonathanlurie commented:
Thanks @nyalldawson for the feedback! I am addressing points by points and your remarks about color coding for types makes a lot of sens. I decided to slightly modify the colors associated to different types, here is my proposal:

**Burgundy rgb(122, 0, 47) for vector types:**
- QgsProcessingParameterPoint
- QgsProcessingParameterGeometry
- QgsProcessingParameterVectorLayer
- QgsProcessingParameterFeatureSource
- QgsProcessingParameterVectorDestination
- QgsProcessingParameterFeatureSink
- QgsProcessingOutputVectorLayer

**Turquoise rgb(0, 180, 180) for raster types:**
- QgsProcessingParameterRasterLayer
- QgsProcessingParameterRasterDestination (should it belong to the file/folder category?)
- QgsProcessingOutputRasterLayer

**Cold gray rgb(137, 150, 171) for other layers:**
- QgsProcessingParameterMultipleLayers
- QgsProcessingParameterMapLayer
- QgsProcessingParameterAnnotationLayer
- QgsProcessingParameterMeshLayer
- QgsProcessingParameterPointCloudLayer
- QgsProcessingOutputVectorTileLayer
- QgsProcessingOutputMapLayer
- QgsProcessingOutputMultipleLayers

**Green rgb(51, 201, 28) for booleans:**
- QgsProcessingParameterBoolean
- QgsProcessingOutputBoolean

**Lavender blue rgb(120, 100, 255) for strings:**
- QgsProcessingParameterCrs
- QgsProcessingParameterString
- QgsProcessingParameterDateTime
- QgsProcessingParameterExpression
- QgsProcessingOutputString
- QgsProcessingOutputHtml

**Dark neutral gray rgb(80, 80, 80) for files/folders:**
- QgsProcessingParameterFile
- QgsProcessingParameterFileDestination
- QgsProcessingParameterFolderDestination
- QgsProcessingParameterPointCloudDestination
- QgsProcessingParameterVectorTileDestination
- QgsProcessingOutputFolder
- QgsProcessingOutputFile


**Blue rgb(34, 157, 214) for numerical values:**
- QgsProcessingParameterMatrix
- QgsProcessingParameterNumber
- QgsProcessingParameterDistance
- QgsProcessingParameterDuration
- QgsProcessingParameterScale
- QgsProcessingParameterRange
- QgsProcessingParameterColor
- QgsProcessingParameterExtent
- QgsProcessingOutputNumber
- QgsProcessingOutputVariant


**Purple rgb(152, 68, 201) for enums:**
- QgsProcessingParameterEnum


**Mid gray rgb(128, 128, 128) for the other QgsProcessingParameterDefinition and QgsProcessingOutputDefinition**QgsProcessingParameterDefinition

Let me know if that's works in terms of grouping?

jonathanlurie committed:
gradient along path when type is fluid

jonathanlurie committed:
shorter gradient

jonathanlurie commented:
As per the values that come from parameters our processing outputs, I suggest to frame the name of the box the value is coming from with `< >` so that the whole field is shorter than in my initial proposal:

![image](https://github.com/user-attachments/assets/fbd1bbb8-3fdd-4bba-92e8-ccae84190bfb)

What do you think?

jonathanlurie committed:
make fields shorter

jonathanlurie committed:
Merge branch 'master' into model-designer-update

jonathanlurie committed:
Merge branch 'master' into model-designer-update

nyalldawson commented:
@jonathanlurie 
> Let me know if that's works in terms of grouping?

Sounds good to me!

nyalldawson commented:
> When types are relatively fluid (from what I understand, it's mostly when a value is captured as a QVariant) I think it's a nice way to convey that the values is "being converted" somewhere in the pipe. Let me know what you think.

I'm ok with that too -- I'm happy for us to experiment here, given that it's not anything which will break a users' workflow if we tweak this in future...

nyalldawson commented:
> As per the values that come from parameters our processing outputs, I suggest to frame the name of the box the value is coming from with < > so that the whole field is shorter than in my initial proposal:

Looks good to me!

jonathanlurie committed:
apply fixes

jonathanlurie commented on src/core/processing/qgsprocessingparameters.h:
```diff
@@ -460,6 +461,17 @@ class CORE_EXPORT QgsProcessingParameterDefinition
 
     virtual ~QgsProcessingParameterDefinition() = default;
 
+    /**
+     * A color to represent the default parameter
+     */
+    virtual QColor getColor() const;
```
Ok, I just moved the colors to the corresponding `QgsProcessingParameterType` and now I just have a `modelColor()` method in the class `QgsProcessingParameterDefinition` that returns the color of the type instance from  processingRegistry.  
I coul'd find a similar logic of attaching the color to the type when it comes to processing outputs, did I miss anything?

nyalldawson replied:
> I coul'd find a similar logic of attaching the color to the type when it comes to processing outputs, did I miss anything?

You didn't miss anything -- there's no equivalent (yet) for outputs :+1:

jonathanlurie committed:
moved param colors to types

nyalldawson commented on src/core/processing/qgsprocessingparametertype.h:
```diff
@@ -171,6 +173,11 @@ class CORE_EXPORT QgsProcessingParameterType
     virtual QList<int> acceptedDataTypes( const QgsProcessingParameterDefinition *parameter ) const;
 
+    /**
+     * A color to represent the default parameter
```
```suggestion
     * Returns the color to use for the parameter type in model designer windows.
     *
     * \since QGIS 3.44
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.h:
```diff
@@ -460,6 +461,17 @@ class CORE_EXPORT QgsProcessingParameterDefinition
     virtual ~QgsProcessingParameterDefinition() = default;
 
+    /**
+     * A color to represent the default parameter
```
```suggestion
     * Returns the color to use for the parameter in model designer windows.
     *
     * The default implementation retrieves the color from the parameter type, see QgsProcessingParameterType::modelColor().
     *
     * \since QGIS 3.44
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.h:
```diff
@@ -460,6 +461,17 @@ class CORE_EXPORT QgsProcessingParameterDefinition
+    virtual QColor getColor() const;
+
+    /**
+     * Get a user friendly string representation of the provided parameter value.
+     * (Meant to be override by child classes to adapt to specific param types)
+     */
+    virtual QString getUserFriendlyValue( QVariant paramValue ) const;
```
This is still unresolved

nyalldawson commented on src/core/processing/qgsprocessingparametertype.cpp:
```diff
@@ -51,3 +56,8 @@ QList<int> QgsProcessingParameterType::acceptedDataTypes( const QgsProcessingPar
+
+QColor QgsProcessingParameterType::modelColor() const
+{
+  return QColor( 128, 128, 128 ); /* mid gray */
```
```suggestion
  return defaultModelColor();
```

nyalldawson commented on src/core/processing/qgsprocessingparametertype.h:
```diff
@@ -34,6 +34,8 @@ class CORE_EXPORT QgsProcessingParameterType
 {
   public:
 
+    static QColor defaultModelColor();
```
add documentation

nyalldawson commented:
@jonathanlurie I've tagged this as frozen now -- it will need to wait for 4.0 as it's too close to 3.44 release to merge changes of this size.  You'll need to update the "since" tags in the documentation accordingly.

jonathanlurie committed:
add geometry type name getter

jonathanlurie committed:
wkt encoded geometries now display geom type in graphic item

jonathanlurie commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -3021,6 +3021,128 @@ QgsProcessingParameterDefinition *QgsProcessingParameterBoolean::clone() const
+      const QString wkt = g.asWkt();
+      return QStringLiteral( "wkt(%1)" ).arg( wkt );
+    }
+    return QStringLiteral( "wkt(null)" );
```
For this, I added the following to `QgsGeometry`:
- `static QString name( Qgis::GeometryType type )`
- `QString typeName() const` (for a given instance)

It uses the `GeometryType` enum under the hood, so there are only a short number of geom types.

jonathanlurie committed:
add translations for invalid data time

jonathanlurie committed:
better crs string

jonathanlurie committed:
rename method to socketColor

jonathanlurie committed:
avoid abbreviations

jonathanlurie committed:
avoid abbreviations 2

jonathanlurie committed:
avoid abbreviations 3

jonathanlurie committed:
Better function naming

jonathanlurie committed:
using static default color

jonathanlurie committed:
updating comments

jonathanlurie committed:
WIP userFriendlString testing

jonathanlurie committed:
fix friendly string for point, add UT

jonathanlurie committed:
adding some more UT

jonathanlurie committed:
Merge branch 'master' into model-designer-update

nyalldawson commented on src/core/processing/qgsprocessingoutputs.h:
```diff
@@ -85,6 +86,13 @@ class CORE_EXPORT QgsProcessingOutputDefinition
+    /**
+    * Returns the color to use for the output in model designer windows.
+    *
+    * \since QGIS 3.44
```
```suggestion
    * \since QGIS 4.0
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -3021,6 +3021,70 @@ QgsProcessingParameterDefinition *QgsProcessingParameterBoolean::clone() const
+
+QString QgsProcessingParameterDefinition::userFriendlyString( const QVariant &value ) const
+{
+  if ( value.userType() == qMetaTypeId<QgsPointXY>() )
```
All these userFriendlyString implementations should start by checking:

```suggestion
   if ( QgsVariantUtils::isNull( value ) )
     return QString();

  if ( value.userType() == qMetaTypeId<QgsPointXY>() )
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -3592,7 +3668,8 @@ QString QgsProcessingParameterExtent::valueAsPythonString( const QVariant &value
            qgsDoubleToString( r.xMaximum() ),
-           qgsDoubleToString( r.yMaximum() ),                                                                                                                             r.crs().authid() );
+           qgsDoubleToString( r.yMaximum() ),
+           r.crs().userFriendlyIdentifier( Qgis::CrsIdentifierType::ShortString ) );
```
```suggestion
           r.crs().authid() );
```

Otherwise it won't be parsed correctly by parameterAsExtent

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -8806,6 +8944,49 @@ QgsProcessingParameterDateTime *QgsProcessingParameterDateTime::fromScriptCode(
+    if ( !dt.isValid() )
+      return QObject::tr( "Invalid datetime" );
+    else
+      return QStringLiteral( "%1-%2-%3T%4:%5:%6" )
```
```suggestion
      return dt.toString( Qt::ISODate );
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -8806,6 +8944,49 @@ QgsProcessingParameterDateTime *QgsProcessingParameterDateTime::fromScriptCode(
+    if ( !dt.isValid() )
+      return QObject::tr( "Invalid date" );
+    else
+      return QStringLiteral( "%1-%2-%3" )
+             .arg( dt.year() )
+             .arg( dt.month() )
+             .arg( dt.day() );
```
```suggestion
      return dt.toString( Qt::ISODate );
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -8806,6 +8944,49 @@ QgsProcessingParameterDateTime *QgsProcessingParameterDateTime::fromScriptCode(
+    if ( !dt.isValid() )
+      return QObject::tr( "Invalid time" );
+    else
+      return QStringLiteral( "%4:%5:%6" )
+             .arg( dt.hour() )
+             .arg( dt.minute() )
+             .arg( dt.second() );
```
```suggestion
      return dt.toString( Qt::ISODate );
```

nyalldawson commented on tests/src/analysis/testqgsprocessing.cpp:
```diff
@@ -3210,6 +3213,8 @@ void TestQgsProcessing::parameterCrs()
   QVERIFY( def->checkValueIsAcceptable( QgsProcessingFeatureSourceDefinition( QgsProperty::fromValue( QVariant::fromValue( r1 ) ) ) ) );
   QVERIFY( def->checkValueIsAcceptable( QgsProcessingOutputLayerDefinition( r1->id() ) ) );
 
+  QCOMPARE( def->userFriendlyString( QVariant( "EPSG:3857" ) ), QgsCoordinateReferenceSystem( QVariant( "EPSG:3857" ).toString() ).userFriendlyIdentifier( Qgis::CrsIdentifierType::ShortString ) );
```
```suggestion
  QCOMPARE( def->userFriendlyString( QVariant( "EPSG:3857" ) ), QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:3857" ) ).userFriendlyIdentifier( Qgis::CrsIdentifierType::ShortString ) );
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -4013,6 +4091,42 @@ QgsProcessingParameterGeometry *QgsProcessingParameterGeometry::fromScriptCode(
+      const QgsReferencedGeometry g = value.value<QgsReferencedGeometry>();
+      if ( !g.isNull() )
+      {
+        return QStringLiteral( "[%1] %2" ).arg( g.crs().userFriendlyIdentifier( Qgis::CrsIdentifierType::ShortString ), g.typeName() );
```
let's flip this order -- the geometry type should come first, then the crs (to match eg extent and point parameter formats)

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -3021,6 +3021,70 @@ QgsProcessingParameterDefinition *QgsProcessingParameterBoolean::clone() const
+  else if ( value.userType() == qMetaTypeId<QgsReferencedPointXY>() )
+  {
+    const QgsReferencedPointXY r = value.value<QgsReferencedPointXY>();
+    return QStringLiteral( "[%1] %2, %3" ).arg(
+             r.crs().authid(),
+             qgsDoubleToString( r.x(), 4 ),
+             qgsDoubleToString( r.y(), 4 )
```
flip order, have crs last

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -3021,6 +3021,70 @@ QgsProcessingParameterDefinition *QgsProcessingParameterBoolean::clone() const
+    if ( !g.isNull() )
+    {
+
+      return QStringLiteral( "[%1] %2" ).arg( g.crs().userFriendlyIdentifier( Qgis::CrsIdentifierType::ShortString ), g.typeName() );
```
flip order ,crs last

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -7811,6 +7935,11 @@ bool QgsProcessingParameterArea::fromVariantMap( const QVariantMap &map )
 
+QString QgsProcessingParameterArea::userFriendlyString( const QVariant &value ) const
+{
+  return QStringLiteral( "%1 %2" ).arg( value.toString(), QgsUnitTypes::toString( defaultUnit() ) );
```
```suggestion
  return QStringLiteral( "%1 %2" ).arg( value.toString(), QgsUnitTypes::toAbbreviatedString( defaultUnit() ) );
```

nyalldawson commented on src/core/processing/qgsprocessingparameters.cpp:
```diff
@@ -7892,6 +8021,10 @@ bool QgsProcessingParameterVolume::fromVariantMap( const QVariantMap &map )
 
+QString QgsProcessingParameterVolume::userFriendlyString( const QVariant &value ) const
+{
+  return QStringLiteral( "%1 %2" ).arg( value.toString(), QgsUnitTypes::toString( defaultUnit() ) );
```
```suggestion
  return QStringLiteral( "%1 %2" ).arg( value.toString(), QgsUnitTypes::toAbbreviatedString( defaultUnit() ) );
```

nyalldawson commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -180,13 +185,32 @@ QgsModelDesignerSocketGraphicItem::QgsModelDesignerSocketGraphicItem( QgsModelCo
-  constexpr float DISPLAY_SIZE = 3.2;
-  painter->drawEllipse( position(), DISPLAY_SIZE, DISPLAY_SIZE );
+  constexpr float DISPLAY_SIZE = 4;
+  float ellipseOffset = 0.4;
```
```suggestion
  constexpr float ELLIPSE_OFFSET = 0.4;
```

Also add a comment on why this is here

nyalldawson commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -197,4 +221,73 @@ void QgsModelDesignerSocketGraphicItem::paint( QPainter *painter, const QStyleOp
+      // Input params
+      case Qt::TopEdge:
+      {
+        QgsProcessingParameterDefinitions params = child->algorithm()->parameterDefinitions();
+
+        if ( mIndex > ( params.length() - 1 ) )
+        {
+          break;
+        }
+
+        const QgsProcessingParameterDefinition *param = params.at( mIndex );
```
We can simplify this a little:
```suggestion
        const QgsProcessingParameterDefinitions params = child->algorithm()->parameterDefinitions();
        const QgsProcessingParameterDefinition *param = params.value( mIndex );
        if ( !param )
          break;
```

nyalldawson commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -197,4 +221,73 @@ void QgsModelDesignerSocketGraphicItem::paint( QPainter *painter, const QStyleOp
+        }
+
+        const QgsProcessingParameterDefinition *param = params.at( mIndex );
+        QString name = param->name();
```
```suggestion
        const QString name = param->name();
```

nyalldawson commented on src/gui/processing/models/qgsmodelgraphicitem.cpp:
```diff
@@ -197,4 +221,73 @@ void QgsModelDesignerSocketGraphicItem::paint( QPainter *painter, const QStyleOp
+        QString name = param->name();
+
+        QgsProcessingModelChildParameterSources paramSources = child->parameterSources().value( name );
+        if ( paramSources.size() == 0 )
```
```suggestion
        if ( paramSources.empty() )
```

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.h:
```diff
@@ -427,6 +437,10 @@ class GUI_EXPORT QgsModelParameterGraphicItem : public QgsModelComponentGraphicI
     void contextMenuEvent( QGraphicsSceneContextMenuEvent *event ) override;
     bool canDeleteComponent() override;
 
+    QString linkedParameterDataType( Qt::Edge /* unused in this implementation because parameters only have a bottom edge */, int index ) const override;
```
move comment to .cpp file

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.h:
```diff
@@ -137,8 +137,12 @@ class GUI_EXPORT QgsModelComponentGraphicItem : public QGraphicsObject
      */
     void setItemRect( QRectF rect );
 
+    virtual QString linkedParameterDataType( Qt::Edge edge, int index ) const;
```
Missing documentation, and "\since"

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.h:
```diff
@@ -137,8 +137,12 @@ class GUI_EXPORT QgsModelComponentGraphicItem : public QGraphicsObject
+
 #ifndef SIP_RUN
 
+    virtual QColor linkColor( Qt::Edge edge, int index ) const;
```
Missing documentation, since

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.h:
```diff
@@ -363,6 +367,11 @@ class GUI_EXPORT QgsModelComponentGraphicItem : public QGraphicsObject
 
+    /**
+     * Returns the fallback color if the parameter or output does not have a specific color.
+     */
```
Add \since

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -563,15 +567,20 @@ void QgsModelComponentGraphicItem::fold( Qt::Edge edge, bool folded )
   else if ( QgsProcessingModelParameter *param = dynamic_cast<QgsProcessingModelParameter *>( mComponent.get() ) )
-    mModel->parameterComponent( param->parameterName() ).setLinksCollapsed( edge, folded );
+  {
+    QgsProcessingModelParameter parameterComponent = mModel->parameterComponent( param->parameterName() );
+    parameterComponent.setLinksCollapsed( edge, folded );
```
This is broken -- you are setting setLinksCollapsed on a copy which you do nothing with. The original code worked because parameterComponent returns a reference.

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -1140,9 +1270,53 @@ QString QgsModelChildAlgorithmGraphicItem::linkPointText( Qt::Edge edge, int ind
+        const QgsProcessingParameterDefinition *param = params.at( index );
+        QString name = param->name();
+        QString title = param->description();
+
```
```suggestion
```

nyalldawson commented on src/gui/processing/models/qgsmodelcomponentgraphicitem.cpp:
```diff
@@ -1140,9 +1270,53 @@ QString QgsModelChildAlgorithmGraphicItem::linkPointText( Qt::Edge edge, int ind
+        QgsProcessingModelChildParameterSources paramSources = child->parameterSources().value( name );
+        QString parameterValueAsString;
+
+        if ( paramSources.size() > 0 )
```
```suggestion
        if ( !paramSources.empty() )
```

nyalldawson commented on src/gui/processing/models/qgsmodelarrowitem.cpp:
```diff
@@ -81,30 +82,31 @@ void QgsModelArrowItem::paint( QPainter *painter, const QStyleOptionGraphicsItem
   {
     case Marker::Circle:
-      painter->drawEllipse( mEndPoint, 3.0, 3.0 );
+      painter->drawEllipse( mEndPoint, 3, 3 );
```
There's no benefit to this change -- the QPainter method takes a qreal argument, so it'll be converted from int anyway.

jonathanlurie committed:
fix after review

jonathanlurie committed:
addresing review

jonathanlurie committed:
removed blank line

jonathanlurie committed:
removed unsused variables

jonathanlurie committed:
removed unsused variables

jonathanlurie committed:
Merge branch 'master' into model-designer-update

jonathanlurie committed:
merge from master

jonathanlurie commented:
Hey @nyalldawson , does it look good to you?

jonathanlurie committed:
Merge branch 'master' into model-designer-update

nyalldawson commented:
Superseded by https://github.com/qgis/QGIS/pull/64170
